# Technical Challenge ! Custiomers and Orders - The lifeblood of any business!

This notebook depends on two data files: customers.scv and orders.csv. These files are hosted in the cloud (or Github), so for simplicity, we are providing you with the code to download and save the two files in a folder called 'sample_data' which lives within this notebook session (not on your local computer nor Google Drive).

![sample_data folder in notebook session](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/images/sample_data_folder.png)

Please execute the first code snippet before moving to the rest of the exercise as this code will import the data from the files for you to be able to execute the rest of the exercise.

Remember as we said in the previous lecture, you have to save this notebook into your Google Drive by going to menu File/Save a Copy In Drive, or clicking in the button "Copy to Drive". This way, you will not lose any work you've done and the file will persist with your latest changes in your Google Drive. Also we recommend you to rename your .ipynb in your Google Drive, so you can easily find it in the future.

In [1]:
import requests

def import_data_files():
  r = requests.get('https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/data/customers.csv')
  with open('./sample_data/customers.csv', 'wb') as f:
    f.write(r.content)

  r = requests.get('https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/data/orders.csv')
  with open('./sample_data/orders.csv', 'wb') as f:
    f.write(r.content)

import_data_files()
print("Customers and orders CSV files have been added './sample_data'")

Customers and orders CSV files have been added './sample_data'


# Exercise 1: Processing Customers data (difficulty medium)

The sample customer data in 'customers.csv' file has just 5 columns: CustomerId, First Name, Last Name, City and State

![Data sample](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/images/customers.png)

We strongly recommend that you complete from the [Prep Course: Intro to Python](https://colab.research.google.com/github/anyoneai/notebooks/blob/main/python3_crash_course.ipynb) the following section:
- Section 7: File I/O (to understand how to read CSV file)
- Section 6: For Loop (to navigate the contents of the CSV file)
- Section 5: Tuples, Lists, and Dictionaries (to manipulate the data of the CSV file)

With this, we hope you can complete this exercise successfully. Although if you want to solve this with libraries or any other way, you are welcome to do it your way.

*Hint:* We advise you to take a look at the data before you start.
**if you want to manually take a look at the data before you start, please see the content of the data [here](https://github.com/anyoneai/notebooks/blob/main/customers_and_orders/data/customers.csv).

*Hint*: There are many ways to do this exercise you can do your own, although here's some help. You can solve this exercise reading and parsing CSV files, structuring data into dictionaries, and using for loops to navigate the contents

*Hint*: Also, keep in mind that data might not be clean and you might have to figure out how to deal with that data from the code, without having to modify the data source.

Below are the 5 questions you'll have to answer to pass the evaluation:

**Question 1:** How many customers are in the file?
(as help, we have added some comments and starter code to help you structure the solution)

In [3]:
import numpy as np
import pandas as pd

ruta = "/content/sample_data/customers.csv"

customer_data = pd.read_csv(ruta)

customer_data

CustomerID    FirstName      LastName         City State
0          6192        Randi    Piedrahita  Canoga Park    CA
1       5100595  Christopher       Abraham   Loma Linda    CA
2       1902451        Elvis      Jauregui  Los Angeles    CA
3       1906377         Jose  Mota-Sanchez       Downey    CA
4       1902406      Scott        McCrorie   Long Beach    CA
..          ...          ...           ...          ...   ...
597     5415365      Trinity     Hall Reif        Poway    CA
598     5415465       Kisona   Mialoto Jr.      Lynwood    CA
599     5415345     Lawrence         Zhang  Temple City    CA
600     5415481         John        Hughes    Round Top    TX
601     5415482      Phillip        Okonma  Los Angeles    CA

[602 rows x 5 columns]

In [7]:
customer_data.shape

(602, 5)

In [6]:
customer_data["CustomerID"].nunique()

471

In [17]:
customer_data.sort_values(by="CustomerID", ascending=True)

CustomerID FirstName     LastName                       City  \
141        5014    Teresa     Ascolese              Santa Barbara   
44         5053       Doe       Harris                Los Angeles   
95         5053       Doe       Harris  Canoga Park                 
94         5053       Doe       Harris            North Hollywood   
443        5068   Maureen  Amato Mayes              Santa Barbara   
..          ...       ...          ...                        ...   
567     5415498     Ayman      Youssef                   Monrovia   
568     5415499  Mariella         Rudi                Los Angeles   
572     5415500     Volha    Kirilenko                  San Diego   
133     5415502  Geovanny     Espinoza                  Claremont   
534     5415503    Alison       Rivera                 South Gate   

             State  
141             CA  
44              CA  
95   CA             
94              CA  
443             CA  
..             ...  
567             CA  
568             CA  
572             CA  
133             CA  
534             CA  

[602 rows x 5 columns]

In [18]:
customer_data["CustomerID"].value_counts()

CustomerID
5463       5
10986      5
1909384    4
4301117    4
700771     4
          ..
5370647    1
5370268    1
5370211    1
5371053    1
5173569    1
Name: count, Length: 471, dtype: int64

### **ANSWER**
As we can see from the analysis made, there are 602 rows of data from customers but `there are 471 unique customers`. There are customers that have more than 1 appearance which can be misleading.

### **RESPUESTA**
Como podemos ver en el análisis realizado, hay 602 filas de datos de clientes, pero `la cantidad de clientes únicos son 471`, ya que hay clientes que tienen más de una instancia en el dataset, lo que puede llevar a pensar que son 602.

**Question 2:** In how many different states do the customers live in?

In [20]:
display(customer_data["State"].nunique())

display(customer_data["State"].unique())

16

array(['CA', 'AZ', 'NV', 'FL', 'CA           ', 'WA', 'NH', 'ID', 'CO',
       'Ca', 'TX', 'NM', 'OR', 'UT', 'MA', 'IN'], dtype=object)

In [25]:
states = customer_data["State"].unique()

states_clean = []

for state in states:
    clean_state = state.lower().strip()
    if clean_state not in states_clean:
        states_clean.append(clean_state)

display(states_clean)
display(len(states_clean))

['ca',
 'az',
 'nv',
 'fl',
 'wa',
 'nh',
 'id',
 'co',
 'tx',
 'nm',
 'or',
 'ut',
 'ma',
 'in']

14

### **ANSWER**
At first glance with the .nunique function we can observe 16 unique/different states but when we observe this values the are some inconsistencies:
- CA is different from Ca and (CA&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;) but they all are referring to California.
- After cleaning this mess we can see `there are 14 unique/different states where customers live in`.

### **RESUMEN**
A primera vista, cuando utilizamos la función .nunique podemos ver que hay 16 diferentes estados donde los clientes residen, pero cuando observamos los valores hay algunas inconsistencias:
- CA es diferente a Ca y (CA&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;) pero todos estos valores hacen referencia a California.
- Despúes de limpiar estos valores obtenemos que `hay 14 estados diferentes donde residen los clientes`.

**Question 3** What is the state with most customers?

In [26]:
customer_data["State"].value_counts()

State
CA               562
NV                 8
AZ                 6
CA                 5
FL                 3
CO                 3
NM                 3
Ca                 2
UT                 2
TX                 2
WA                 1
NH                 1
ID                 1
OR                 1
MA                 1
IN                 1
Name: count, dtype: int64

### **ANSWER**
We can clearly see the difference between California (even without the messy CA values) and the rest of the states.
`California is the state with most customers`.

### **RESPUESTA**
Podemos observar con claridad que California le lleva una amplia diferencia al resto de los estados (incluso sin los valores incorrectos de CA).
`California es el estado con más clientes`.


**Question 4** What is the state with the least customers?

In [33]:
customer_data["State"].value_counts().sort_values(ascending=True)

State
ID                 1
OR                 1
IN                 1
MA                 1
WA                 1
NH                 1
Ca                 2
UT                 2
TX                 2
CO                 3
NM                 3
FL                 3
CA                 5
AZ                 6
NV                 8
CA               562
Name: count, dtype: int64

### **ANSWER**
There are 6 states that have a unique customer:
- `Idaho`
- `Oregon`
- `Indiana`
- `Massachusetts`
- `Washington`
- `New Hampshire`

This 6 states are the answer.

### **RESPUESTA**
Hay 6 estados que tienen un único cliente:
- `Idaho`
- `Oregon`
- `Indiana`
- `Massachusetts`
- `Washington`
- `New Hampshire`

Por lo que estos estados satisfacen la respuesta.

**Question 5:** What is the most common last name?

In [59]:
customer_data["LastName"].value_counts()

LastName
Smith       8
Doggett     5
Gomez       5
Zambrana    5
Johnson     4
           ..
Merle       1
Valdez      1
Muggy       1
Persky      1
Tan         1
Name: count, Length: 424, dtype: int64

In [58]:
last_names_count = {}

for last_name in customer_data["LastName"]:
    clean_last_name = last_name.strip().lower()
    if clean_last_name not in last_names_count:
        last_names_count[clean_last_name] = 1
    else:
        last_names_count[clean_last_name] += 1

last_name_df = pd.DataFrame(last_names_count.items(), columns=["LastName", "Count"])
last_name_df.sort_values(by="Count", ascending=False)

LastName  Count
7       smith      8
213   doggett      5
64      gomez      5
96   zambrana      5
55    johnson      4
..        ...    ...
401     merle      1
400    valdez      1
11     persky      1
8         tan      1
6       klink      1

[423 rows x 2 columns]

### **ANSWER**
We used the .value_counts function to get an idea of the possible answers but since we do not know the inconsistencies in the data we check each row individually and normalized. We achieve the same result for the answer:
`Smith is the most common last name`.

### **RESPUESTA**
Utilizamos la función .value_counts para tener una idea de las posibles respuestas pero como pueden existir inconsistencias en los datos también lo buscamos luego de normalizar las filas.
`Smith es el apellido más común/frecuente`.

# Exercise 2: Processing Orders data (difficulty high)

The second sample files contains orders placed by customers from the first file. Be careful, this file has many rows and you most likely should not print the contents of the whole file.

The file contains the following columns: CustomerID, OrderID, Date, OrderTotal, ProductName, Price

![Data sample](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/images/orders.png)

*Hint:* We advise you to take a look at the data before you start.
**if you want to manually take a look at the data before you start, please see the content of the data [here](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/data/orders.csv).

*Hint*: There are many ways to do this exercise you can do your own, although here's some help. You can solve this exercise reading and parsing CSV files, structuring data into dictionaries, and using for loops to navigate the contents

*Hint*: Also, the data is not clean and you will have to figure out how to deal with that data from the code, without having to modify the data source.



**Question #1:** How many unique orders are in the orders.csv file?

**Question #2:** What is the average number of items per order (rounded to two decimal places)?

**Question #3:** What is the highest number of items per order?

**Question #4:** What is the number of orders placed in October 2021?

**Question #5:** Which customer spent the most amount of money in 2021?

**Question #6:** Historically, what is the best month for sales?

Once you get your answers, remember to go back to the course and introduce them in the multiple choice quiz

In [61]:
ruta_2 = "/content/sample_data/orders.csv"

orders_data = pd.read_csv(ruta_2)

orders_data.head()

CustomerID  OrderID                     Date  OrderTotal ProductName  Price
0        8091  7742581  2021-07-26 14:40:10.783        95.0         Z03   90.0
1      902139  7742778  2021-08-08 05:01:21.120        60.0    0844 A/C   60.0
2     2300266  7742593  2021-07-27 11:00:16.020       185.0         M07   90.0
3     2300266  7742593  2021-07-27 11:00:16.020       185.0         M09   90.0
4     5173013  7742609  2021-07-28 14:26:13.930       165.0        0324  160.0

In [63]:
orders_data.shape

(29294, 6)

**Question #1:** How many unique orders are in the orders.csv file?

In [62]:
orders_data["OrderID"].nunique()

16672

### **ANSWER**
`There are 16672 unique orders in the file`.

### **RESPUESTA**
`Hay 16672 órdenes únicas en el archivo`.

**Question #2:** What is the average number of items per order (rounded to two decimal places)?

In [66]:
round(orders_data.groupby("OrderID").size().mean(), 2)

np.float64(1.76)

### **ANSWER**
The average number of items per order is: `1.76`.

### **RESPUESTA**
El promedio de items por orden es: `1.76`.

**Question #3:** What is the highest number of items per order?

In [67]:
orders_data.groupby("OrderID").size().sort_values(ascending=False)

OrderID
5802666    35
7001632    30
7553706    30
7411097    30
7005266    28
           ..
7791825     1
7791824     1
7791823     1
7791822     1
7791821     1
Length: 16672, dtype: int64

### **ANSWER**
The highest number of items in an order is: `35`.

### **RESPUESTA**
La máxima cantidad de items en una orden es: `35`.

**Question #4:** What is the number of orders placed in October 2021?

In [74]:
orders_data[(orders_data["Date"] >= "2021-10") & (orders_data["Date"] < "2021-11")]["OrderID"].nunique()

267

### **ANSWER**
`There are 267 unique orders in October 2021`.

### **RESPUESTA**
`Hay 267 ordenes únicas en el mes de Octubre de 2021`.

**Question #5:** Which customer spent the most amount of money in 2021?

In [91]:
data_2021 = orders_data[orders_data["Date"].between("2021-01", "2022-01")]
data_2021 = data_2021.drop_duplicates(subset="OrderID")

data_2021.groupby("CustomerID")["OrderTotal"].sum().sort_values(ascending=False)

CustomerID
5172443    7755.0
1907160    5500.0
5464       5480.0
5100984    5225.0
4300438    5017.0
            ...  
5370602     105.0
5370758     105.0
5370884     105.0
5370422      95.0
5370388      90.0
Name: OrderTotal, Length: 342, dtype: float64

In [95]:
data_2021[data_2021["CustomerID"] == 5172443]

CustomerID  OrderID                     Date  OrderTotal ProductName  \
1406      5172443  7694774  2021-03-23 12:35:53.483       485.0        0116   
4340      5172443  7741602  2021-06-28 14:46:22.897       710.0        0983   
4574      5172443  7742654  2021-07-30 17:04:39.080       935.0        1922   
7749      5172443  7694893  2021-04-05 14:22:51.813       250.0        0113   
8063      5172443  7695807  2021-05-05 10:56:15.877       735.0        0110   
8130      5172443  7740599  2021-05-21 16:57:25.000       700.0        0983   
8892      5172443  7743529  2021-08-20 16:39:42.690      1125.0        0983   
9288      5172443  7788604  2021-10-01 12:57:56.307       935.0        1920   
10231     5172443  7789483  2021-10-28 10:43:26.493       235.0        0913   
10232     5172443  7789484  2021-10-28 10:47:38.840       710.0        1920   
26564     5172443  7790462  2021-12-06 15:12:11.783       935.0        1920   

       Price  
1406   250.0  
4340   225.0  
4574   225.0  
7749   250.0  
8063   250.0  
8130   225.0  
8892   225.0  
9288   225.0  
10231  225.0  
10232  225.0  
26564  225.0

In [92]:
customer_data[customer_data["CustomerID"] == 5172443]

CustomerID FirstName LastName         City State
171     5172443   Brandon    Divas  Los Angeles    CA

### **ANSWER**
`The customer with ID 5172443 spent the most amount of money in 2021`.

### **RESPUESTA**
`El cliente con ID 5172443 fue el que más gastó en 2021`.

**Question #6:** Historically, what is the best month for sales?

In [83]:
orders_data_copy = orders_data.copy()

orders_data_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29294 entries, 0 to 29293
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   CustomerID   29294 non-null  int64  
 1   OrderID      29294 non-null  int64  
 2   Date         27152 non-null  object 
 3   OrderTotal   29294 non-null  float64
 4   ProductName  29294 non-null  object 
 5   Price        29294 non-null  float64
dtypes: float64(2), int64(2), object(2)
memory usage: 1.3+ MB


In [96]:
orders_data_copy["Date"] = pd.to_datetime(orders_data_copy["Date"])
orders_data_copy["Month"] = orders_data_copy["Date"].dt.month
orders_data_copy_unique = orders_data_copy.drop_duplicates(subset="OrderID")
orders_data_copy.groupby("Month")["OrderTotal"].sum().sort_values(ascending=False)

Month
1.0     858727.0
2.0     831150.0
5.0     798050.0
6.0     753265.0
9.0     747421.0
11.0    745817.0
10.0    597649.0
7.0     585495.0
8.0     575493.0
4.0     560733.0
3.0     556715.0
12.0    510948.0
Name: OrderTotal, dtype: float64

### **ANSWER**
January is the best month for sales with 858727 in OrderTotal. Close to it is February with 831150.

### **RESPUESTA**
Enero es el mejor mes de ventas, con 858727 en OrderTotal. Cerca del mismo está Febrero con 831150.

# Finished!

Hope this was not too difficult and slicing and dicing the datasets was some fun. Now head on back to the course and provide the answers to the questions from this exercise.